packages/react/src/hooks/useController.ts

```diff
@@ -0,0 +1,83 @@
+import { OverlayEventDetail } from '@ionic/core';
+import { useMemo, useRef } from 'react';
+
+import { attachProps } from '../components/utils';
+
+import { HookOverlayOptions } from './HookOverlayOptions';
+
+interface OverlayBase extends HTMLElement {
+  present: () => Promise<void>;
+  dismiss: (data?: any, role?: string | undefined) => Promise<boolean>;
+}
+
+export function useController<
+  OptionsType,
+  OverlayType extends OverlayBase
+>(
+  displayName: string,
+  controller: { create: (options: OptionsType) => Promise<OverlayType> }
+) {
+  const overlayRef = useRef<OverlayType>();
+  const didDismissEventName = useMemo(
+    () => `on${displayName}DidDismiss`,
+    [displayName]
+  );
+  const didPresentEventName = useMemo(
+    () => `on${displayName}DidPresent`,
+    [displayName]
+  );
+  const willDismissEventName = useMemo(
+    () => `on${displayName}WillDismiss`,
+    [displayName]
+  );
+  const willPresentEventName = useMemo(
+    () => `on${displayName}WillPresent`,
+    [displayName]
+  );
+
+  const present = async (options: OptionsType & HookOverlayOptions) => {
+    if (overlayRef.current) {
+      return;
+    }
+    const {
+      onDidDismiss,
+      onWillDismiss,
+      onDidPresent,
+      onWillPresent,
+      ...rest
+    } = options;
+
+    overlayRef.current = await controller.create({
+      ...(rest as any),
+    });
+
+    attachProps(overlayRef.current, {
+      [didDismissEventName]: handleDismiss,
+      [didPresentEventName]: (e: CustomEvent) =>
+        onDidPresent && onDidPresent(e),
+      [willDismissEventName]: (e: CustomEvent) =>
+        onWillDismiss && onWillDismiss(e),
+      [willPresentEventName]: (e: CustomEvent) =>
+        onWillPresent && onWillPresent(e),
+    });
+
+    overlayRef.current.present();
+
+    function handleDismiss(event: CustomEvent<OverlayEventDetail<any>>) {
```

Comment:
Small thing, but can you make this a const arrow function, and place it directly after another const? Just a minifying trick